go-controller/pkg/ovn/address_set_sync.go

```diff
@@ -0,0 +1,348 @@
+				return nil, fmt.Errorf("found LogicalRouterPolicy that doesn't belong to the %s router, can't update", ovntypes.OVNClusterRouter)
+			}
+			ops, err = libovsdbops.CreateOrUpdateLogicalRouterPoliciesOps(syncer.nbClient, ops, ovntypes.OVNClusterRouter, lrp)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get update LRPs ops: %v", err)
+			}
+		}
+	}
+	return
+}
+
+// getAddrSetUpdateInfo adds db ops to update address set and objects that reference it
+func (syncer *addressSetsSyncer) getAddrSetUpdateInfo(as *nbdb.AddressSet) (*updateAddrSetInfo, error) {
+	oldName, ipSuffix := truncateSuffixFromAddressSet(as.ExternalIDs["name"])
+	// oldName may be empty if address set doesn't have ExternalID set
+	acls, qoses, lrps, dbIDs, err := syncer.getReferencingObjsAndNewDbIDs(as.Name, oldName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get new dbIDs for address set %s: %v", oldName, err)
+	}
+	updateAddrSet := &updateAddrSetInfo{acls, qoses, lrps, as, nil}
+
+	if oldName == "" {
+		klog.Infof("external_ids->name missing, stale address set %s", as.Name)
+		return updateAddrSet, nil
+	}
+	if ipSuffix == "" {
+		klog.Infof("Found stale address set %s without ip family suffix and empty ips list", oldName)
+		return updateAddrSet, nil
+	}
+	nbdbASv4, nbdbASv6 := addressset.GetDbObjsForAS(dbIDs, nil)
+	var nbdbAS *nbdb.AddressSet
+	if ipSuffix == ipv4AddressSetSuffix {
+		nbdbAS = nbdbASv4
+	} else {
+		nbdbAS = nbdbASv6
+	}
+	// since we need to update addressSet.Name, which is an index and not listed in getNonZeroAddressSetMutableFields,
+	// we copy existing addressSet, update it address set needed, and replace (delete and create) existing address set with the updated
+	newAS := as.DeepCopy()
+	// reset UUID
+	newAS.UUID = ""
+	// update address set Name
+	newAS.Name = nbdbAS.Name
+	// remove old "name" ExternalID
+	delete(newAS.ExternalIDs, "name")
+	// Insert new externalIDs generated by address set factory
+	for key, value := range nbdbAS.ExternalIDs {
+		newAS.ExternalIDs[key] = value
+	}
+	updateAddrSet.newAddrSet = newAS
+	return updateAddrSet, nil
+}
+
+func (syncer *addressSetsSyncer) SyncAddressSets() error {
+	// stale address sets don't have controller ID
+	p := libovsdbops.GetNoOwnerPredicate[*nbdb.AddressSet]()
+	addrSetList, err := libovsdbops.FindAddressSetsWithPredicate(syncer.nbClient, p)
+	if err != nil {
+		return fmt.Errorf("failed to find stale address sets: %v", err)
+	}
+	updatedCount := 0
+	defer func() {
+		klog.Infof("SyncAddressSets handled %d of %d stale address sets, %d of them were ignored",
+			updatedCount-syncer.ignoredAddressSets, len(addrSetList), syncer.ignoredAddressSets)
+	}()
+	lastBatch := len(addrSetList) / syncer.txnBatchSize
+	if len(addrSetList)%syncer.txnBatchSize > 0 {
+		lastBatch += 1
+	}
+	for batchIdx := 0; batchIdx < lastBatch; batchIdx++ {
+		maxIdx := (batchIdx + 1) * syncer.txnBatchSize
+		if len(addrSetList) < maxIdx {
+			maxIdx = len(addrSetList)
+		}
```

Comment:
Suggestion for a simpler batching where `txnBatchSize == 0` means no batching

```
i := 0
for i < len(addrSetList) {
    for j := 1; j <= txnBatchSize && i < len(addrSetList); j++, i++ {
    	updateInfo, err := syncer.getAddrSetUpdateInfo(addrSetList[j])
	if err != nil {
		return err
	}
	addrSetInfos = append(addrSetInfos, updateInfo)
    }
    // get ops
    // transact
}

```

Reply:
this exact loop won't work for no batching, but I updated it and added a unit test for disabled batching